packages/fx-core/src/core/middleware/aadManifestMigration.ts

```diff
@@ -0,0 +1,272 @@
+import {
+  RequiredResourceAccess,
+  AADManifest,
+} from "../../plugins/resource/aad/interfaces/AADManifest";
+import { Constants } from "../../plugins/resource/aad/constants";
+import { PluginNames } from "../../plugins/solution/fx-solution/constants";
+import * as os from "os";
+
+const upgradeButton = "Upgrade";
+let userCancelFlag = false;
+const backupFolder = ".backup";
+
+interface Permission {
+  resource: string;
+  delegated: string[];
+  application: string[];
+}
+
+export const AadManifestMigrationMW: Middleware = async (
+  ctx: CoreHookContext,
+  next: NextFunction
+) => {
+  if (await needMigrateToArmAndMultiEnv(ctx)) {
+    next();
+  } else if (await needConsolidateLocalRemote(ctx)) {
+    next();
+  } else if ((await needMigrateToAadManifest(ctx)) && checkMethod(ctx)) {
+    sendTelemetryEvent(Component.core, TelemetryEvent.ProjectMigratorNotificationStart);
+    const res = await TOOLS?.ui.showMessage(
+      "warn",
+      getLocalizedString("core.aadManifestMigration.Message"),
+      true,
+      upgradeButton
+    );
+    const answer = res?.isOk() ? res.value : undefined;
+    if (!answer || answer != upgradeButton) {
+      sendTelemetryEvent(Component.core, TelemetryEvent.ProjectAadManifestMigrationNotification, {
+        [TelemetryProperty.Status]: ProjectMigratorStatus.Cancel,
+      });
+      ctx.result = err(AadManifestMigrationCanceledError());
+      outputCancelMessage(ctx);
+      return;
+    }
+
+    try {
+      await migrate(ctx);
+      await next();
+    } catch (error) {
+      sendTelemetryErrorEvent(
+        Component.core,
+        TelemetryEvent.ProjectAadManifestMigrationError,
+        assembleError(error, CoreSource)
+      );
+      throw error;
+    }
+  } else {
+    await next();
+  }
+};
+
+async function needMigrateToAadManifest(ctx: CoreHookContext): Promise<boolean> {
+  try {
+    if (!isConfigUnifyEnabled() || !isAadManifestEnabled()) {
+      return false;
+    }
+
+    const inputs = ctx.arguments[ctx.arguments.length - 1] as Inputs;
+    if (!inputs.projectPath) {
+      return false;
+    }
+    const fxExist = await fs.pathExists(path.join(inputs.projectPath as string, ".fx"));
+    if (!fxExist) {
+      return false;
+    }
+
+    const aadManifestTemplateExist = await fs.pathExists(
+      path.join(inputs.projectPath as string, "templates", "appPackage", "aad.template.json")
+    );
+
+    if (aadManifestTemplateExist) {
+      return false;
+    }
+
+    const permissionFileExist = await fs.pathExists(
+      path.join(inputs.projectPath as string, "permissions.json")
+    );
+    const projectSettingsJson = await fs.readJson(
+      path.join(inputs.projectPath as string, ".fx", "configs", "projectSettings.json")
+    );
+    const aadPluginIsActive = projectSettingsJson.solutionSettings.activeResourcePlugins.includes(
+      PluginNames.AAD
+    );
+
+    if (!aadPluginIsActive || !permissionFileExist) {
```

Comment:
Since we won't delete permission file, maybe migration will be called several times?

Reply:
we move this file to backup folder at last